kernel/work.c

```diff
@@ -0,0 +1,1030 @@
+	 * point because the lock was held.  If this is being invoked by a
+	 * preemptible thread then yield.
+	 */
+	if ((ret > 0) && (k_is_preempt_thread() != 0)) {
+		k_yield();
+	}
+
+
+	return ret;
+}
+
+/* Flush the work item if necessary.
+ *
+ * Flushing is necessary only if the work is either queued or running.
+ *
+ * Invoked with work lock held by key.
+ * Sleeps.
+ *
+ * @param work the work item that is to be flushed
+ * @param flusher state used to synchronize the flush
+ * @param key the key under which the work lock is held.
+ *
+ * @retval true if work is queued or running.  If this happens the
+ * caller must take the flusher semaphore after releasing the lock.
+ *
+ * @retval false otherwise.  No wait required.
+ */
+static bool work_flush_locked(struct k_work *work,
+			      struct z_work_flusher *flusher)
+{
+	struct k_work_q *queue = NULL;
+	bool need_flush = (flags_get(&work->flags)
+			   & (K_WORK_QUEUED | K_WORK_RUNNING)) != 0;
+
+	if (need_flush) {
+		queue = work->queue;
+		__ASSERT_NO_MSG(queue != NULL);
+
+		queue_flusher_locked(queue, work, flusher);
+		notify_queue_locked(queue);
+	}
+
+	return need_flush;
+}
+
+bool k_work_flush(struct k_work *work,
+		  struct k_work_sync *sync)
+{
+	__ASSERT_NO_MSG(work != NULL);
+	__ASSERT_NO_MSG(!flag_test(&work->flags, K_WORK_DELAYABLE_BIT));
+	__ASSERT_NO_MSG(!k_is_in_isr());
+	__ASSERT_NO_MSG(sync != NULL);
+
+	struct z_work_flusher *flusher = &sync->flusher;
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	bool need_flush = work_flush_locked(work, flusher);
+
+	k_spin_unlock(&lock, key);
+
+	/* If necessary wait until the flusher item completes */
+	if (need_flush) {
+		k_sem_take(&flusher->sem, K_FOREVER);
+	}
+
+	return need_flush;
+}
+
+/* Execute the non-waiting steps necessary to cancel a work item.
+ *
+ * Invoked with work lock held.
+ *
+ * @param work the work item to be canceled.
+ *
+ * @retval true if we need to wait for the work item to finish canceling
+ * @retval false if the work item is idle
+ *
+ * @return k_busy_wait() captured under lock
+ */
+static int cancel_async_locked(struct k_work *work)
+{
+	/* If we haven't already started canceling, do it now. */
+	if (!flag_test(&work->flags, K_WORK_CANCELING_BIT)) {
+		/* Remove it from the queue, if it's queued. */
+		queue_remove_locked(work->queue, work);
+	}
+
+	/* If it's still busy after it's been dequeued, then flag it
+	 * as canceling.
+	 */
+	int ret = work_busy_get_locked(work);
+
+	if (ret != 0) {
+		flag_set(&work->flags, K_WORK_CANCELING_BIT);
+		ret = work_busy_get_locked(work);
+	}
+
+	return ret;
+}
+
+/* Complete cancellation necessary, release work lock, and wait if
+ * necessary.
+ *
+ * Invoked with work lock held by key.
+ * Sleeps.
+ *
+ * @param work work that is being canceled
+ * @param canceller state used to synchronize the cancellation
+ * @param key used by work lock
+ *
+ * @retval true iff the work was still active on entry.  The caller
```

Comment:
iff == if and only if; expanded the text.

Reply:
You are right. I have never seen it being used. It may be just me but I expect this is not widely known unless somebody studied logic in English. Anyway - I will recall it, thanks.
https://en.wikipedia.org/wiki/If_and_only_if

I had similar situation with double space used by people from UK. I have not seen it used outside.